File: code/ex106.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int i;
    public Map<Integer, Integer> val2idx = new HashMap<>();
    
    public TreeNode build(int[] postorder, int lo, int hi) {
        if (lo > hi)    return null;
        
        TreeNode root = new TreeNode(postorder[i]);
        int idx = val2idx.get(postorder[i]);
        i--;
        
        root.right = build(postorder, idx + 1, hi);
        root.left = build(postorder, lo, idx - 1);
        
        return root;
    }   
    
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        for (int i = 0; i < inorder.length; i++)
            val2idx.put(inorder[i], i);
        i = postorder.length - 1;
        return build(postorder, 0, postorder.length - 1);   
    }
}

File: code/ex106.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        int p = postorder.size() - 1;
        return build(inorder, postorder, 0, inorder.size() - 1, p);
    }
    
private:
    TreeNode* build(vector<int>& inorder, vector<int>& postorder, int ilo, int ihi, int& p) {
        if (ilo > ihi)    return NULL;
        TreeNode* root = new TreeNode(postorder[p--]);
        int pivot = ilo;
        while (inorder[pivot] != root->val)    pivot++;
        root->right = build(inorder, postorder, pivot + 1, ihi, p);
        root->left = build(inorder, postorder, ilo, pivot - 1, p);
        return root;
    }
};